File: libraries/ndf/ndf1_infcb.f
      SUBROUTINE NDF1_INFCB( STATUS )
*+
*  Name:
*     NDF1_INFCB

*  Purpose:
*     Initialise the NDF_ Format Conversion Block.

*  Language:
*     Starlink Fortran 77

*  Invocation:
*     CALL NDF1_INFCB( STATUS )

*  Description:
*     The routine initialises the NDF_ Format Conversion Block (FCB)
*     which maintains information about the foreign data formats which
*     should be recognised by the NDF_ library and which may be
*     converted to and from native NDF format. This information is
*     obtained by translating appropriate environment variables.
*
*     This routine only executes once. If called again after a
*     successful initial invocation it will return without further
*     action.

*  Arguments:
*     STATUS = INTEGER (Given and Returned)
*        The global status.

*  Copyright:
*     Copyright (C) 1993 Science & Engineering Research Council

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK, RAL)
*     {enter_new_authors_here}

*  History:
*     12-OCT-1993 (RFWS):
*        Original version.
*     12-NOV-1993 (RFWS):
*        Added initialisation of NDF extension name list.
*     15-APR-1994 (RFWS):
*        Added initialisation of the anonymous converter array.
*     31-OCT-1994 (RFWS):
*        Removed initialisation of NDF extension list information (now
*        performed separately for each foreign file).
*     {enter_further_changes_here}

*  Bugs:
*     {note_any_bugs_here}

*-

*  Type Definitions:
      IMPLICIT NONE              ! No implicit typing

*  Global Constants:
      INCLUDE 'SAE_PAR'          ! Standard SAE constants
      INCLUDE 'NDF_CONST'        ! NDF_ private constants
      INCLUDE 'NDF_ERR'          ! NDF_ error codes

*  Global Variables:
      INCLUDE 'NDF_FCB'          ! NDF_ Format Conversion Block
*        FCB_FEX1( 2 * NDF__MXFMT ) = INTEGER (Write)
*           Starting positions of each foreign format file name
*           extension in the FCB_FMT string.
*        FCB_FEX2( 2 * NDF__MXFMT ) = INTEGER (Write)
*           Ending positions of each foreign format file name extension
*           in the FCB_FMT string.
*        FCB_FMT = CHARACTER * ( 2 * NDF__SZFMT ) (Write)
*           String to contain list of input and output foreign format
*           specifications.
*        FCB_FMT1( 2 * NDF__MXFMT ) = INTEGER (Write)
*           Starting positions of each foreign format name in the
*           FCB_FMT string.
*        FCB_FMT2( 2 * NDF__MXFMT ) = INTEGER (Write)
*           Ending positions of each foreign format name in the FCB_FMT
*           string.
*        FCB_NIN = INTEGER (Write)
*           Number of foreign formats to recognise on input.
*        FCB_NOUT = INTEGER (Write)
*           Number of foreign formats to recognise on output.

*  External References:
      INTEGER CHR_LEN            ! Significant length of a string
      LOGICAL CHR_SIMLR          ! Case insensitive string comparison

*  Status:
      INTEGER STATUS             ! Global status

*  Local Variables:
      INTEGER E1                 ! First extension character position
      INTEGER E2                 ! Last extension character position
      INTEGER F                  ! First character position
      INTEGER F1                 ! First format character position
      INTEGER F2                 ! Last format character position
      INTEGER IFMT               ! Loop counter for formats
      INTEGER L                  ! Last character position
      INTEGER LVAL               ! Length of environment variable value
      LOGICAL DEF                ! Environment variable defined?
      LOGICAL FIRST              ! First invocation of this routine?
      LOGICAL SDF                ! File extension is reserved (.sdf)?

      SAVE FIRST                 ! Remember if previously invoked

*  Local Data:
      DATA FIRST / .TRUE. /      ! Routine initially not yet invoked

*.

*  Check inherited global status.
      IF ( STATUS .NE. SAI__OK ) RETURN

*  Only perform initialisation on the first invocation of this routine.
      IF ( FIRST ) THEN

*  Input formats.
*  =============
*  Translate the environment variable NDF_FORMATS_IN to obtain a list
*  of foreign data formats which are to be recognised on input. Put the
*  translation into the first half of the FCB format list string. Note
*  if there is no translation.
         CALL NDF1_GTENV( 'NDF_FORMATS_IN', DEF,
     :                    FCB_FMT( : NDF__SZFMT ), LVAL, STATUS )

*  If no translation was found, then no foreign data formats will be
*  recognised on input. Otherwise, parse the format list to split it
*  into separate format fields, storing the field positions temporarily
*  at the start of the format name field position list.
         FCB_NIN = 0
         IF ( ( STATUS .EQ. SAI__OK ) .AND. DEF ) THEN
            CALL NDF1_PSFFL( FCB_FMT( : NDF__SZFMT ), NDF__MXFMT,
     :                       FCB_FMT1( 1 ), FCB_FMT2( 1 ), FCB_NIN,
     :                       STATUS )

*  If OK, then loop to split each field into a foreign format name and
*  its associated file type extension string (which will be used to
*  identify files with that format).
            IF ( STATUS .EQ. SAI__OK ) THEN
               DO 1 IFMT = 1, FCB_NIN
                  F = FCB_FMT1( IFMT )
                  L = FCB_FMT2( IFMT )
                  CALL NDF1_PSFMT( FCB_FMT( F : L ), F1, F2, E1, E2,
     :                             STATUS )

*  Quit looping if an error occurs.
                  IF ( STATUS .NE. SAI__OK ) GO TO 2

*  Correct the returned field positions for the starting position used
*  and store the results at the start of the FCB field position lists.
                  FCB_FMT1( IFMT ) = F1 + F - 1
                  FCB_FMT2( IFMT ) = F2 + F - 1
                  FCB_FEX1( IFMT ) = E1 + F - 1
                  FCB_FEX2( IFMT ) = E2 + F - 1

*  Convert all the foreign format names to upper case and check that the
*  reserved data format name "NDF" has not been specified. Report an
*  error if it has.
                  F = FCB_FMT1( IFMT )
                  L = FCB_FMT2( IFMT )
                  CALL CHR_UCASE( FCB_FMT( F : L ) )
                  IF ( FCB_FMT( F : L ) .EQ. 'NDF' ) THEN
                     STATUS = NDF__FMTIN
                     CALL ERR_REP( 'NDF1_INFCB_NDF1',
     :                    'The format name ''NDF'' is reserved and ' //
     :                    'may not be used to identify a foreign ' //
     :                    'data format.', STATUS )
                     GO TO 2
                  END IF

*  Check whether the specified file extension will clash with that used
*  for native format NDF files (be case insensitive if necessary).
                  F = FCB_FEX1( IFMT )
                  L = FCB_FEX2( IFMT )
                  CALL NDF1_CMPFL( FCB_FMT( F : L ), '.sdf', SDF,
     :                             STATUS )
                  IF ( STATUS .NE. SAI__OK ) GO TO 2

*  Report an error if the file extension clashes.
                  IF ( SDF ) THEN
                     STATUS = NDF__FMTIN
                     CALL MSG_SETC( 'SDF', FCB_FMT( F : L ) )
                     CALL ERR_REP( 'NDF1_INFCB_SDF1',
     :                    'The file extension ''^SDF'' is reserved ' //
     :                    'and may not be used to identify a ' //
     :                    'foreign format data file.', STATUS )
                     GO TO 2
                  END IF
 1             CONTINUE
 2             CONTINUE
            END IF

*  If an error occurred, then report contextual information.
            IF ( STATUS .NE. SAI__OK ) THEN
               CALL ERR_REP( 'NDF1_INFCB_IN',
     : 'Error occurred while reading the NDF_FORMATS_IN foreign ' //
     : 'data format list (possible bad environment variable setting).',
     :                       STATUS )
            END IF
         END IF

*  Output formats.
*  ==============
*  In the same way, translate the environment variable NDF_FORMATS_OUT
*  to obtain a list of foreign data formats which are to be recognised
*  on output. Put the translation into the second half of the FCB
*  format list string. Note if there is no translation.
         CALL NDF1_GTENV( 'NDF_FORMATS_OUT', DEF,
     :                    FCB_FMT( NDF__SZFMT + 1 : ), LVAL, STATUS )

*  If no translation was found, then no foreign data formats will be
*  recognised on output. Otherwise, parse the format list to split it
*  into separate format fields, storing the field positions temporarily
*  at the end of the format name field position list.
         FCB_NOUT = 0
         IF ( ( STATUS .EQ. SAI__OK ) .AND. DEF ) THEN
            CALL NDF1_PSFFL( FCB_FMT( NDF__SZFMT + 1 : ), NDF__MXFMT,
     :                       FCB_FMT1( NDF__MXFMT + 1 ),
     :                       FCB_FMT2( NDF__MXFMT + 1 ), FCB_NOUT,
     :                       STATUS )

*  If OK, then loop to split each field into a foreign format name and
*  its associated file type extension string (which will be used to
*  identify files with that format).
            IF ( STATUS .EQ. SAI__OK ) THEN
               DO 3 IFMT = NDF__MXFMT + 1, NDF__MXFMT + FCB_NOUT
                  F = FCB_FMT1( IFMT ) + NDF__SZFMT
                  L = FCB_FMT2( IFMT ) + NDF__SZFMT

*  If the format specified was simply '*' or '.', then accept it
*  without further checks and set both the format name and the file
*  extension position pointers to identify it.
                  IF ( ( FCB_FMT( F : L ) .EQ. '*' ) .OR.
     :                 ( FCB_FMT( F : L ) .EQ. '.' ) ) THEN
                     FCB_FMT1( IFMT ) = F
                     FCB_FMT2( IFMT ) = L
                     FCB_FEX1( IFMT ) = F
                     FCB_FEX2( IFMT ) = L

*  Otherwise parse the format specification.
                  ELSE
                     CALL NDF1_PSFMT( FCB_FMT( F : L ), F1, F2, E1, E2,
     :                                STATUS )

*  Quit looping if an error occurs.
                     IF ( STATUS .NE. SAI__OK ) GO TO 4

*  Correct the returned field positions for the starting position used
*  and store the results at the end of the FCB field position lists.
                     FCB_FMT1( IFMT ) = F1 + F - 1
                     FCB_FMT2( IFMT ) = F2 + F - 1
                     FCB_FEX1( IFMT ) = E1 + F - 1
                     FCB_FEX2( IFMT ) = E2 + F - 1

*  Convert all the foreign format names to upper case and check that the
*  reserved data format name "NDF" has not been specified. Report an
*  error if it has.
                     F = FCB_FMT1( IFMT )
                     L = FCB_FMT2( IFMT )
                     CALL CHR_UCASE( FCB_FMT( F : L ) )
                     IF ( FCB_FMT( F : L ) .EQ. 'NDF' ) THEN
                        STATUS = NDF__FMTIN
                        CALL ERR_REP( 'NDF1_INFCB_NDF2',
     :                       'The format name ''NDF'' is reserved ' //
     :                       'and may not be used to identify a ' //
     :                       'foreign data format.', STATUS )
                        GO TO 4
                     END IF

*  Check whether the specified file extension will clash with that used
*  for native format NDF files (be case insensitive if necessary).
                     F = FCB_FEX1( IFMT )
                     L = FCB_FEX2( IFMT )
                     CALL NDF1_CMPFL( FCB_FMT( F : L ), '.sdf', SDF,
     :                                STATUS )
                     IF ( STATUS .NE. SAI__OK ) GO TO 4

*  Report an error if the file extension clashes.
                     IF ( SDF ) THEN
                        STATUS = NDF__FMTIN
                        CALL MSG_SETC( 'SDF', FCB_FMT( F : L ) )
                        CALL ERR_REP( 'NDF1_INFCB_SDF2',
     :                       'The file extension ''^SDF'' is ' //
     :                       'reserved and may not be used to ' //
     :                       'identify a foreign format data file.',
     :                       STATUS )
                        GO TO 4
                     END IF
                  END IF
 3             CONTINUE
 4             CONTINUE
            END IF

*  If an error occurred, then report contextual information.
            IF ( STATUS .NE. SAI__OK ) THEN
               CALL ERR_REP( 'NDF1_INFCB_OUT',
     : 'Error occurred while reading the NDF_FORMATS_OUT foreign ' //
     : 'data format list (possible bad environment variable setting).',
     :                       STATUS )
            END IF
         END IF

*  Note if the first invocation of this routine has completed
*  successfully.
         IF ( STATUS .EQ. SAI__OK ) FIRST = .FALSE.

*  Call error tracing routine if necessary.
         IF ( STATUS .NE. SAI__OK ) CALL NDF1_TRACE( 'NDF1_INFCB',
     :                                               STATUS )
      END IF

      END

File: libraries/ndf/ndf1Infcb.c
#include <stdlib.h>
#include <string.h>
#include "sae_par.h"
#include "ndf1.h"
#include "ndf_err.h"
#include "mers.h"
#include "ndf1_types.h"

void ndf1Infcb( int *status ){
/*
*+
*  Name:
*     ndf1Infcb

*  Purpose:
*     Initialise the NDF_ Format Conversion Block.

*  Synopsis:
*     void ndf1Infcb( int *status )

*  Description:
*     This function initialises the NDF_ Format Conversion Block (FCB)
*     which holds information about the foreign data formats which
*     should be recognised by the NDF_ library and which may be converted
*     to and from native NDF format. This information is obtained by
*     translating appropriate environment variables.
*
*     This function only executes once, during global initialisation. It
*     should not be called again.

*  Parameters:
*     *status
*        The global status.

*  Copyright:
*     Copyright (C) 2018 East Asian Observatory
*     All rights reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or modify
*     it under the terms of the GNU General Public License as published by
*     the Free Software Foundation; either version 2 of the License, or (at
*     your option) any later version.
*
*     This program is distributed in the hope that it will be useful,but
*     WITHOUT ANY WARRANTY; without even the implied warranty of
*     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
*     General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     RFWS: R."f". Warren-Smith (STARLINK)
*     DSB: David S. Berry (EAO)

*  History:
*     3-APR-2019 (DSB):
*        Original version, based on equivalent Fortran function by RFWS.

*-
*/

/* Local Variables: */
   NdfFCB *fcb;          /* Pointer to next FCB structure */
   char *fmt;            /* Pointer to next extracted format description */
   char fmts[ NDF__SZFMT + 1 ]; /* Comma-separated list of formats */
   int def;              /* Environment variable defined? */
   int f;                /* First character position */
   int ifcb;             /* Loop counter for formats */
   int l;                /* Last character position */
   int nc;               /* Length of string */
   int nfmt;             /* Number of formats in list */
   int sdf;              /* File extension is reserved (.sdf)? */
   size_t e1;            /* First extension character position */
   size_t e2;            /* Last extension character position */
   size_t f1;            /* First format character position */
   size_t f2;            /* Last format character position */
   size_t ibeg[ NDF__MXFMT ]; /* Index of start of each format specifier */
   size_t iend[ NDF__MXFMT ]; /* Index of end of each format specifier */
   size_t lval;          /* Length of environment variable value */



/* Check inherited global status. */
   if( *status != SAI__OK ) return;

/* Ensure the current thread has exclusive access to the tuning
   parameters. */
   NDF__TCB_LOCK_MUTEX;

/* Input formats.
   =============
   Translate the environment variable NDF_FORMATS_IN to obtain a list
   of foreign data formats which are to be recognised on input. Note
   if there is no translation. */
   Ndf_TCB_forin = 0;
   ndf1Gtenv( "NDF_FORMATS_IN", &def, fmts, sizeof( fmts ), &lval, status );

/* If no translation was found, then no foreign data formats will be
   recognised on input. Otherwise, parse the format list to split it
   into separate format fields, storing the field positions temporarily
   at the start of the format name field position list. */
   if( *status == SAI__OK && def ) {
      ndf1Psffl( fmts, NDF__MXFMT, ibeg, iend, &nfmt, status );

/* If OK, then loop to split each field into a foreign format name and
   its associated file type extension string (which will be used to
   identify files with that format). */
      if( *status == SAI__OK ) {
         fmt = NULL;
         for( ifcb = 0; ifcb < nfmt; ifcb++ ){
            f = ibeg[ ifcb ];
            l = iend[ ifcb ];

            fmt = ndf1Strip( fmt, fmts, f, l, NULL, NULL, status );
            ndf1Psfmt( fmt, &f1, &f2, &e1, &e2, status );

/* Get a pointer to a new FCB structure in which to store the information. */
            fcb = ndf1Ffs( NDF__FCBTYPE, status );

/* Quit looping if an error occurs. */
            if( *status != SAI__OK ) break;

/* Indicate we have one or more foreign output formats. */
            Ndf_TCB_forin = 1;

/* Extract the name into the FCB structure. */
            nc = f2 - f1 + 1;
            if( nc > NDF__SZFMN ) nc = NDF__SZFMN;
            memcpy( fcb->name, fmt + f1, nc );
            fcb->name[ nc ] = 0;

/* Extract the extension type into the FCB structure. */
            nc = e2 - e1 + 1;
            if( nc > NDF__SZFMX ) nc = NDF__SZFMX;
            memcpy( fcb->ext, fmt + e1, nc );
            fcb->ext[ nc ] = 0;

/* Indicate this is an input format. */
            fcb->infmt = 1;

/* Convert the foreign format name to upper case and check that the
   reserved data format name "NDF" has not been specified. Report an
   error if it has. */
            astChrCase( NULL, fcb->name, 1, 0 );
            if( !strcmp( fcb->name, "NDF" ) && *status == SAI__OK ) {
               *status = NDF__FMTIN;
               errRep( " ", "The format name 'NDF' is reserved and may "
                       "not be used to identify a foreign data format.",
                       status );
               break;
            }

/* Check whether the specified file extension will clash with that used
   for native format NDF files (be case insensitive if necessary). */
            ndf1Cmpfl( fcb->ext, 1, 0, ".sdf", &sdf, status );
            if( *status != SAI__OK ) break;

/* Report an error if the file extension clashes. */
            if( sdf ) {
               *status = NDF__FMTIN;
               msgSetc( "SDF", fcb->ext );
               errRep( " ", "The file extension '^SDF' is reserved and "
                       "may not be used to identify a foreign format "
                       "data file.", status );
               break;
            }
         }

/* Free the string holding the current format. */
         fmt = astFree( fmt );
      }

/* If an error occurred, then report contextual information. */
      if( *status != SAI__OK ) {
         errRep( " ", "Error occurred while reading the NDF_FORMATS_IN "
                 "foreign data format list (possible bad environment "
                 "variable setting).", status );
      }
   }

/* Output formats.
   ==============
   In the same way, translate the environment variable NDF_FORMATS_OUT
   to obtain a list of foreign data formats which are to be recognised
   on output. Put the translation into the second half of the FCB
   format list string. Note if there is no translation. */
   Ndf_TCB_forout = 0;
   ndf1Gtenv( "NDF_FORMATS_OUT", &def, fmts, sizeof( fmts ), &lval, status );

/* If no translation was found, then no foreign data formats will be
   recognised on output. Otherwise, parse the format list to split it
   into separate format fields, storing the field positions temporarily
   at the end of the format name field position list. */
   if( ( *status == SAI__OK ) && def ) {
      ndf1Psffl( fmts, NDF__MXFMT, ibeg, iend, &nfmt, status );

/* If OK, then loop to split each field into a foreign format name and
   its associated file type extension string (which will be used to
   identify files with that format). */
      if( *status == SAI__OK ) {
         fmt = NULL;
         for( ifcb = 0; ifcb < nfmt; ifcb++ ){
            f = ibeg[ ifcb ];
            l = iend[ ifcb ];
            fmt = ndf1Strip( fmt, fmts, f, l, NULL, NULL, status );

/* Get a pointer to a new FCB structur in which to store the information. */
            fcb = ndf1Ffs( NDF__FCBTYPE, status );

/* Indicate this is an output format. */
            if( fcb ) {
               fcb->infmt = 0;

/* If the format specified was simply "*" or ".", then accept it
   without further checks and set both the format name and the file
   extension position pointers to identify it. */
               if( !strcmp( fmt, "*" ) || !strcmp( fmt, "." ) ) {
                  strcpy( fcb->name, fmt );
                  strcpy( fcb->ext, fmt );

/* Otherwise parse the format specification. */
               } else {
                  ndf1Psfmt( fmt, &f1, &f2, &e1, &e2, status );

/* Quit looping if an error occurs. */
                  if( *status != SAI__OK ) break;

/* Indicate we have one or more foreign output formats. */
                  Ndf_TCB_forout = 1;

/* Extract the name into the FCB structure. */
                  nc = f2 - f1 + 1;
                  if( nc > NDF__SZFMN ) nc = NDF__SZFMN;
                  memcpy( fcb->name, fmt + f1, nc );
                  fcb->name[ nc ] = 0;

/* Extract the extension type into the FCB structure. */
                  nc = e2 - e1 + 1;
                  if( nc > NDF__SZFMX ) nc = NDF__SZFMX;
                  memcpy( fcb->ext, fmt + e1, nc );
                  fcb->ext[ nc ] = 0;

/* Convert the foreign format name to upper case and check that the
   reserved data format name "NDF" has not been specified. Report an
   error if it has. */
                  astChrCase( NULL, fcb->name, 1, 0 );
                  if( !strcmp( fcb->name, "NDF" ) && *status == SAI__OK ) {
                     *status = NDF__FMTIN;
                     errRep( " ", "The format name 'NDF' is reserved and may "
                             "not be used to identify a foreign data format.",
                             status );
                     goto L4;
                  }

/* Check whether the specified file extension will clash with that used
   for native format NDF files (be case insensitive if necessary). */
                  ndf1Cmpfl( fcb->ext, 1, 0, ".sdf", &sdf, status );
                  if( *status != SAI__OK ) goto L4;

/* Report an error if the file extension clashes. */
                  if( sdf ) {
                     *status = NDF__FMTIN;
                     msgSetc( "SDF", fcb->ext );
                     errRep( " ", "The file extension '^SDF' is reserved "
                             "and may not be used to identify a foreign "
                             "format data file.", status );
                     goto L4;
                  }
               }
            }
         }
L4:

/* Free the string holding the current format. */
         fmt = astFree( fmt );
      }

/* If an error occurred, then report contextual information. */
      if( *status != SAI__OK ) {
         errRep( " ", "Error occurred while reading the NDF_FORMATS_OUT "
                 "foreign data format list (possible bad environment "
                 "variable setting).", status );
      }
   }

/* Allow other threads to access the tuning parameters. */
   NDF__TCB_UNLOCK_MUTEX;

/* Call error tracing function if necessary. */
   if( *status != SAI__OK ) ndf1Trace( "ndf1Infcb", status );

}

